### PR TITLE
ENH: parse categoricals in read_csv

### LINKED ISSUE #10153: Support categorical variables with CSVs
It would be nice to be able to read CSVs with categorical variables using read_csv's dtype parameter instead of casting the columns after the fact.

### PR DESCRIPTION
**Pull request by chris-b1**
Closes #10153 (at least partly)

Adds the ability to directly parse a `Categorical` through the `dtype` parameter to `read_csv`.  Currently just uses whatever is there as the categories, a possible enhancement would be to allow and enforce user-specified categories, through not quite sure what the api would be.

This only parses string categories - originally I had an impl that did type inference on the categories, but it added a lot of complication without much benefit, now the recommendation in the docs is to convert after parsing.

Here's an example timing.  For reasonably sparse data, a slightly worse than 2x speedup is what I'm typically seeing, along with much better memory usage.

``` python
group1 = ['aaaaa', 'bbbbb', 'cccccc', 'ddddddd', 'eeeeeeee']

df = pd.DataFrame({'a': np.random.choice(group1, 10000000).astype('object'),
                   'b': np.random.choice(group1, 10000000).astype('object'),
                   'c': np.random.choice(group1, 10000000).astype('object')})
df.to_csv('strings.csv', index=False)


In [14]: %timeit pd.read_csv('strings.csv').apply(pd.Categorical)
1 loops, best of 3: 6.66 s per loop

In [13]: %timeit pd.read_csv('strings.csv', dtype='category')
1 loop, best of 3: 3.68 s per loop
```

### CODE REVIEW

File: pandas/parser.pyx [hunk 1]
```diff
@@ -217,6 +221,24 @@ cdef extern from "parser/tokenizer.h":
...
+# XXX
+# this is a hack - in order to make the inference
+# functions generic (converting either data directly
+# from the parser or from a passed in hash table)
+# we add an "optional" parameter via fused type, that can either
+# be the hash table to parse, or an integer, which is used
+# as a sentinel to specialize the function for reading
+# from the parser.
+
+# This is to avoid duplicating a bunch of code or
```
Comment (jreback) [inline pandas/parser.pyx]: why is this needed?

File: pandas/parser.pyx [hunk 2]
```diff
@@ -1151,8 +1181,15 @@ cdef class TextReader:
             # unicode variable width
             return self._string_convert(i, start, end, na_filter,
                                         na_hashset)
-
-
+        # is this comparison good enough?
+        elif dtype == '|O08':
```
Comment (jreback) [inline pandas/parser.pyx]: what is this ``dtype == ...`` ?

File: pandas/parser.pyx [hunk 3]
```diff
@@ -1468,6 +1505,97 @@ cdef _string_box_decode(parser_t *parser, int col,
 
     return result, na_count
 
+@cython.boundscheck(False)
+cdef _categorical_convert(parser_t *parser, int col,
```
Comment (jreback) [inline pandas/parser.pyx]: I am not sure I would do it this way. I would first parse the actual dtype of the column (e.g. into integer, datetimes, whatever), then just convert directly to a categorical. Not sure why this needs a whole factorizing step which is very duplicative.

File: pandas/parser.pyx (hunk 1 above)
Comment (chris-b1) [inline pandas/parser.pyx]: The type inference functions (e.g. `_try_double`) read data straight from the parser -  the `COLITER_NEXT(it,word)` stuff.  I wanted to also be able to pass in a hash table to that function, i.e. my categories, and this was my workaround.

File: pandas/parser.pyx (hunk 3 above)
Comment (chris-b1) [inline pandas/parser.pyx]: The thought was it'd be more performant - you already have to hash each element to compare to na, so why create the uniques at that point, and use the much smaller set for type inference.

Comment (jreback) [inline pandas/parser.pyx]: ahh, yes that makes sense. Actually the same strategy for caching in date parsing. Ok, but instead of rewriting the factorization, can you simply call it with the strings? much simpler

File: pandas/parser.pyx (hunk 2 above)
Comment (chris-b1) [inline pandas/parser.pyx]: This is `dtype.str`, e.g.

```python
In [8]: np.dtype('int64').str
Out[8]: '<i8'

In [10]: CategoricalDtype().str
Out[10]: '|O08'

```

File: pandas/parser.pyx (hunk 3 above)
Comment (chris-b1) [inline pandas/parser.pyx]: The string inference paths factorize and box in the same function, so would have to refactor, but that could work.

File: pandas/parser.pyx (hunk 2 above)
Comment (jreback) [inline pandas/parser.pyx]: hmm, don't you have the original (user passed dtype) ? e.g. ``category``? this is very problematic to do this

File: pandas/parser.pyx (hunk 3 above)
Comment (jreback) [inline pandas/parser.pyx]: just trying to reduce complexity

File: pandas/parser.pyx (hunk 2 above)
Comment (chris-b1) [inline pandas/parser.pyx]: you don't at this point (caller does `.str`), but doesn't seem to be any good reason why, basically the original code - https://github.com/pydata/pandas/commit/3c77f2929c9a81fc8ceeab92d1b75fe41e273474

File: pandas/parser.pyx (hunk 1 above)
Comment (jreback) [inline pandas/parser.pyx]: you would instead set this in parser state and pass directly to the function when its called (e.g. maybe a list off the cols that you want to do hash tracking)

Comment (chris-b1) [inline pandas/parser.pyx]: if I put it in the parser state then checks like [this](https://github.com/pydata/pandas/pull/13406/files#diff-10cd69d969114cd15ba284e3e5da6e22R1720) would get moved to runtime, which I was trying to avoid, though to be fair I haven't profiled to see what it costs.

File: pandas/parser.pyx (hunk 3 above)
Comment (chris-b1) [inline pandas/parser.pyx]: The hash tables make this hard to re-use, the string functions have a table mapping bytes->object, where I need bytes->int here.  I could kill a lot of duplicated code with some templating, ala #13399, but I'm not sure that actually reduces complexity, because then there's another codegen step.

Comment (jreback) [inline pandas/parser.pyx]: what are you needing to code gen? that's not a if deal to add another function and it's at compile time

Comment (chris-b1) [inline pandas/parser.pyx]: Well for instance, instead of my fused type hack, just generate two version of `_try_double`, one that reads off the parser, one that takes an input.  Similarly, lines 1533-1548 below are basically generic and repeated in like 7 functions. (`_try_double`, `_try_int64`, `_string_box_decode`, etc)

Comment (jreback) [inline pandas/parser.pyx]: I c. Maybe it would be better to make a more generic version of these functions (maybe separate PR)? then you can use them in either case.

**Comment by chris-b1** [comment]
@jreback - see if you like this any better.  I am still using a fused type with inference functions, that's really the best I can think of to make them generic, still very open to suggestions though. 

I did add Cython's type specialization syntax, e.g. `_try_double[use_parser_data](self.parser, ...)` which at least make the intent pretty explicit.  It's not that far off something like [tag based dispatch](http://www.boost.org/community/generic_programming.html#tag_dispatching) in C++, though not sure if it's really idiomatic Cython.

**Comment by jreback** [comment]
@chris-b1 like to update this. but also remove the fused type hack. see what you can do.

**Comment by chris-b1** [comment]
@jreback - just getting back to this.  How would you feel about only parsing string/object `Categoricals`?  That would significantly clean up the code but still cover the biggest use-case, and I could just show in the docs how to convert to numeric after the fact if needed.

**Comment by jreback** [comment]
@chris-b1 yes that sounds reasonable

**Comment by chris-b1** [comment]
@jreback - I've updated this to remove the type inference, if you want to have another look.

**Comment by jreback** [comment]
looks pretty - will have to have a more detailed look

can u post perf numbers? (or update top)
with new impl

**Comment by chris-b1** [comment]
Updated performance at the top, although it's basically the same as it was before.

File: doc/source/io.rst [hunk 4]
```diff
@@ -440,6 +440,41 @@ individual columns:
     Specifying ``dtype`` with ``engine`` other than 'c' raises a
     ``ValueError``.
 
+Specifying Categorical dtype
```
Comment (jreback) [inline doc/source/io.rst]: add a reference to this (and refernce from v0.19.0)

File: doc/source/io.rst [hunk 5]
```diff
@@ -440,6 +440,41 @@ individual columns:
...
+''''''''''''''''''''''''''''
+
+.. versionadded:: 0.19.0
+
+`Categorical` columns can be parsed directly by specifying `dtype='category'`
```
Comment (jreback) [inline doc/source/io.rst]: use double-backticks

File: doc/source/io.rst [hunk 6]
```diff
@@ -440,6 +440,41 @@ individual columns:
...
+
+.. ipython :: python
```
Comment (jreback) [inline doc/source/io.rst]: this has to be:

``.. ipython:: python`` (no space)

File: doc/source/io.rst [hunk 7]
```diff
@@ -440,6 +440,41 @@ individual columns:
...
+
+   The resulting categories will always be parsed as string (object dtype).
+   Numeric categories can be converted using the :func:`pd.to_numeric` function.
+
+   .. ipython :: python
+
+      df = pd.read_csv(StringIO(data), dtype='category')
+      df.dtypes
+      df['col3']
+      df['col3'].cat.categories = pd.to_numeric(df['col3'].cat.categories)
```
Comment (jreback) [inline doc/source/io.rst]: this is not idiomatic, we don't assign directly to ``.categories``, instead use ``.set_categories``

```
In [17]: df = pd.DataFrame({'A' : pd.Categorical(['1','2','1'])})

In [18]: df.A
Out[18]:
0    1
1    2
2    1
Name: A, dtype: category
Categories (2, object): [1, 2]

In [19]: df['A'] = df['A'].cat.set_categories(pd.to_numeric(df['A'].cat.categories), rename=True)

In [20]: df.A
Out[20]:
0    1
1    2
2    1
Name: A, dtype: category
Categories (2, int64): [1, 2]
```

I *could* see adding a ``.apply_categories`` accessor, so this would become:

``df['A'] = df['A'].cat.apply_categories(pd.to_numeric)`` which could do the apply-set-with-rename idiom

Comment (jreback) [inline doc/source/io.rst]: actually on 2nd thought, this is ok. assigning directly to ``cat.categories`` is fine

Comment (jreback) [inline doc/source/io.rst]: cc @jorisvandenbossche 
cc @janschulz

File: doc/source/whatsnew/v0.19.0.txt [hunk 8]
```diff
@@ -163,6 +169,44 @@ New behaviour:
 
    In [2]: pd.read_csv(StringIO(data), names=names)
 
+
+.. _whatsnew_0190.enhancements.read_csv_categorical:
+
+:func:`read_csv` supports parsing `Categorical` directly
```
Comment (jreback) [inline doc/source/whatsnew/v0.19.0.txt]: add this to highlites as well (single line that references to here)

Comment (jreback) [inline doc/source/whatsnew/v0.19.0.txt]: use double-back ticks

File: doc/source/whatsnew/v0.19.0.txt [hunk 9]
```diff
@@ -163,6 +169,44 @@ New behaviour:
...
+Individual columns can be parsed as a `Categorical` using a dict specification
+
+.. ipython :: python
+
+   pd.read_csv(StringIO(data), dtype={'col1': 'category'}).dtypes
+
+.. note::
+
+   The resulting categories will always be parsed as string (object dtype).
+   Numeric categories can be converted using the :func:`pd.to_numeric` function.
```
Comment (jreback) [inline doc/source/whatsnew/v0.19.0.txt]: say that IF they are numeric you can post-process them (e.g you can actually do datetimes with categorical as well)

File: pandas/parser.pyx [hunk 10]
```diff
@@ -470,15 +479,10 @@ cdef class TextReader:
...
 
         if isinstance(dtype, dict):
-            conv = {}
-            for k in dtype:
-                v = dtype[k]
-                if isinstance(v, basestring):
-                    v = np.dtype(v)
-                conv[k] = v
-            dtype = conv
+            dtype = {k: pandas_dtype(dtype[k])
```
Comment (jreback) [inline pandas/parser.pyx]: good!

File: pandas/parser.pyx [hunk 11]
```diff
@@ -1134,90 +1132,80 @@ cdef class TextReader:
...
+        elif is_categorical_dtype(dtype):
+            codes, cats, na_count = _categorical_convert(self.parser, i, start,
+                                                         end, na_filter, na_hashset,
+                                                         self.c_encoding)
+            return Categorical(codes, categories=cats, ordered=False,
+                               fastpath=True), na_count
+        elif is_object_dtype(dtype):
             return self._string_convert(i, start, end, na_filter,
                                         na_hashset)
         else:
```
Comment (jreback) [inline pandas/parser.pyx]: this should be an elif

File: pandas/parser.pyx [hunk 12]
```diff
@@ -1134,90 +1132,80 @@ cdef class TextReader:
...
-            if dtype[1] == 'M':
+            if is_datetime64_dtype(dtype):
                  raise TypeError("the dtype %s is not supported for parsing, "
                                  "pass this column using parse_dates instead" % dtype)
             raise TypeError("the dtype %s is not supported for parsing" % dtype)
```
Comment (jreback) [inline pandas/parser.pyx]: make this the else

File: pandas/parser.pyx [hunk 13]
```diff
@@ -1134,90 +1132,80 @@ cdef class TextReader:
...
             return result, na_count
 
-        elif dtype[1] == 'b':
+        elif is_bool_dtype(dtype):
             result, na_count = _try_bool_flex(self.parser, i, start, end,
                                               na_filter, na_hashset,
                                               self.true_set, self.false_set)
             return result, na_count
-        elif dtype[1] == 'c':
-            raise NotImplementedError("the dtype %s is not supported for parsing" % dtype)
```
Comment (jreback) [inline pandas/parser.pyx]: we have a test for not-implemented dtypes? or just always raise TypeError (which is better anyhow)

File: pandas/parser.pyx [hunk 14]
```diff
@@ -1134,90 +1132,80 @@ cdef class TextReader:
...
+        cdef StringPath path = _string_path(self.c_encoding)
+
+        if path == UTF8:
+            return _string_box_utf8(self.parser, i, start, end, na_filter,
+                                    na_hashset)
+        elif path == ENCODED:
+            return _string_box_decode(self.parser, i, start, end,
+                                      na_filter, na_hashset, self.c_encoding)
+        elif path == CSTRING:
+            return _string_box_factorize(self.parser, i, start, end,
```
Comment (jreback) [inline pandas/parser.pyx]: could this actually  be NULL?

**Comment by jreback** [comment]
can you test when chunking with dtype='category', for with and w/o different categories in each chunk (it looks to work), but pls test.

File: doc/source/io.rst (hunk 7 above)
Comment (janschulz) [inline doc/source/io.rst]: IMO assigning to categoires is ok, we even have it in the docs: https://github.com/pydata/pandas/blob/master/pandas/core/categorical.py#L1857

Code is here and checks for same length as old cats: https://github.com/pydata/pandas/blob/master/pandas/core/categorical.py#L546-L570

`set_categories` ([source](https://github.com/pydata/pandas/blob/master/pandas/core/categorical.py#L623-L691)) was designed to do more:

> This method can be used to perform more than one action of adding,
        removing, and reordering simultaneously and is therefore faster than
        performing the individual steps via the more specialised methods.
        On the other hand this methods does not do checks (e.g., whether the
        old categories are included in the new categories on a reorder), which
        can result in surprising changes, for example when using special string
        dtypes on python3, which does not considers a S1 string equal to a
        single char python string.

File: pandas/parser.pyx (hunk 13 above)
Comment (chris-b1) [inline pandas/parser.pyx]: This wasn't tested, so I switched it to `TypeError` and added a test.

File: pandas/parser.pyx (hunk 14 above)
Comment (chris-b1) [inline pandas/parser.pyx]: No - the encoding may be NULL, but `_string_path` will always return a valid enum member.

**Comment by chris-b1** [comment]
@jreback - updated for your comments

File: doc/source/io.rst [hunk 15]
```diff
@@ -440,6 +440,45 @@ individual columns:
...
+.. _io.categorical:
+
+Specifying Categorical dtype
+''''''''''''''''''''''''''''
+
+.. versionadded:: 0.19.0
+
+``Categorical`` columns can be parsed directly by specifying ``dtype='category'``
+
+.. ipython :: python
```
Comment (jorisvandenbossche) [inline doc/source/io.rst]: no space after ipython

Comment (jorisvandenbossche) [inline doc/source/io.rst]: otherwise it is interpreted as a comment I think (rst ..)

File: doc/source/io.rst [hunk 16]
```diff
@@ -440,6 +440,45 @@ individual columns:
...
+
+.. ipython:: python
+
+   pd.read_csv(StringIO(data), dtype={'col1': 'category'}).dtypes
+
+.. note::
+
+   The resulting categories will always be parsed as string (object dtype).
+   If the categories are numeric they can be converted using the
+   :func:`pd.to_numeric` function, or as appropriate, another converter
```
Comment (jorisvandenbossche) [inline doc/source/io.rst]: no 'pd' here. You can do either pandas.to_numeric, or just to_numeric (sphinx knows the current module is pandas, so you can leave that out, but sphinx does not know about the pd abbreviation)

File: doc/source/whatsnew/v0.19.0.txt [hunk 17]
```diff
@@ -222,6 +231,46 @@ New behaviour:
 
    In [2]: pd.read_csv(StringIO(data), names=names)
 
+
+.. _whatsnew_0190.enhancements.read_csv_categorical:
+
+:func:`read_csv` supports parsing ``Categorical`` directly
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v0.19.0.txt]: the underline needs to be at least as long as the title (otherwise you get warnings in the doc build, not that bad, but like to keep the number of warnings as low as possible so the important warnings do not get lost)

File: doc/source/whatsnew/v0.19.0.txt [hunk 18]
```diff
@@ -222,6 +231,46 @@ New behaviour:
...
+.. ipython:: python
+
+   pd.read_csv(StringIO(data), dtype={'col1': 'category'}).dtypes
+
+.. note::
+
+   The resulting categories will always be parsed as string (object dtype).
+   If the categories are numeric they can be converted using the
+   :func:`pd.to_numeric` function, or as appropriate, another converter
+   such as :func:`pd.to_datetime`.
```
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v0.19.0.txt]: same comment about pd here as above

File: pandas/io/tests/parser/c_parser_only.py [hunk 19]
```diff
@@ -184,6 +190,93 @@ def test_pass_dtype(self):
...
+        tm.assert_frame_equal(actual, expected)
+
+        # unsorted
+        data = """a,b,c
+1,b,3.4
+1,b,3.4
+2,a,4.5"""
+        expected = pd.DataFrame({'a': Categorical(['1', '1', '2']),
+                                 'b': Categorical.from_codes([0, 0, 1],
+                                                             ['b', 'a']),
```
Comment (jorisvandenbossche) [inline pandas/io/tests/parser/c_parser_only.py]: I would personally expect the categories to be sorted lexigocraphically, just like you would get with:

```
In [79]: pd.Categorical(['b', 'b', 'a'])
Out[79]:
[b, b, a]
Categories (2, object): [a, b]
```

File: doc/source/io.rst (hunk 15 above)
Comment (chris-b1) [inline doc/source/io.rst]: This actually did work when I built the docs locally, but I also had meant to change it, for consistency if nothing else.

File: pandas/io/tests/parser/c_parser_only.py (hunk 19 above)
Comment (chris-b1) [inline pandas/io/tests/parser/c_parser_only.py]: In `union_categorical` we had decided to order categories as they appear in the groups, but I agree sorting in this case would be more consistent.

Comment (jreback) [inline pandas/io/tests/parser/c_parser_only.py]: maybe add as an option to union_categorical itself

File: doc/source/io.rst (hunk 15 above)
Comment (jorisvandenbossche) [inline doc/source/io.rst]: ah, good to know, then the github rst renderer is more picky than sphinx itself (as it is rendered as a comment)

Comment (jreback) [inline doc/source/io.rst]: you could add this to the lint check 

just grep thru rst and validate the pattern

File: pandas/io/tests/parser/c_parser_only.py [hunk 20]
```diff
@@ -184,6 +190,93 @@ def test_pass_dtype(self):
...
+        tm.assert_frame_equal(actual, expected)
+
+    def test_categorical_dtype_encoding(self):
+        # GH 10153
+        pth = tm.get_data_path('unicode_series.csv')
+        encoding = 'latin-1'
+        expected = self.read_csv(pth, header=None, encoding=encoding)
+        actual = self.read_csv(pth, header=None, encoding=encoding,
+                               dtype={1: 'category'})
+        actual[1] = actual[1].astype(object)
```
Comment (jorisvandenbossche) [inline pandas/io/tests/parser/c_parser_only.py]: @chris-b1 Is it possible you did switch expected and actual? (it seems more logical to me to test that reading in with an encoding actually returns category dtypes, rather than converting that to object and then comparing, as in this way, you didn't check that it was category dtype)

Comment (chris-b1) [inline pandas/io/tests/parser/c_parser_only.py]: I was just being lazy here - related to your other comment, because the current impl doesn't sort categories, it was easier to cast to object (mainly I wanted to test that the decode path was working).  I'll update this too.

**Comment by chris-b1** [comment]
@jreback, @jorisvandenbossche  - this is updated and rebased if you'd like to take another look - I think I have all the previous comments addressed.  In particular the `categories` of the parsed `Categorical` are now sorted.

File: pandas/parser.pyx [hunk 21]
```diff
@@ -1136,90 +1136,88 @@ cdef class TextReader:
...
-        elif dtype[1] == 'O':
+        elif is_categorical_dtype(dtype):
+            codes, cats, na_count = _categorical_convert(self.parser, i, start,
+                                                         end, na_filter, na_hashset,
+                                                         self.c_encoding)
+            # sort categories and recode if necessary
+            cats = Index(cats)
+            if not cats.is_monotonic_increasing:
+                unsorted = cats.copy()
+                cats = cats.sort_values()
```
Comment (sinhrks) [inline pandas/parser.pyx]: You can use ``return_indexer=True`` to get indexer also.

**Comment by jreback** [comment]
thanks @chris-b1 

as usual, check out the docs when built and issue a followup if needed!
